packages/core/core-flows/src/order/workflows/create-return.ts

```diff
@@ -0,0 +1,326 @@
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Cannot apply return reason category with id ${hasInvalidReasons.join(
+        ", "
+      )} to order with id ${orderId}.`
+    )
+  }
+}
+
+function prepareShippingMethodData({
+  orderId,
+  inputShippingOption,
+  returnShippingOption,
+}: {
+  orderId: string
+  inputShippingOption: OrderWorkflow.CreateOrderReturnWorkflowInput["return_shipping"]
+  returnShippingOption: ShippingOptionDTO & WithCalculatedPrice
+}) {
+  const obj: CreateOrderShippingMethodDTO = {
+    name: returnShippingOption.name,
+    order_id: orderId,
+    shipping_option_id: returnShippingOption.id,
+    amount: 0,
+    data: {},
+    // Computed later in the flow
+    tax_lines: [],
+    adjustments: [],
+  }
+
+  if (isDefined(inputShippingOption.price) && inputShippingOption.price >= 0) {
+    obj.amount = inputShippingOption.price
+  } else {
+    if (returnShippingOption.price_type === "calculated") {
+      // TODO: retrieve calculated price and assign to amount
+    } else {
+      obj.amount = returnShippingOption.calculated_price.calculated_amount
+    }
+  }
+
+  return obj
+}
+
+function validateCustomRefundAmount({
+  order,
+  refundAmount,
+}: {
+  order: Pick<OrderDTO, "item_total">
+  refundAmount?: number
+}) {
+  // validate that the refund prop input is less than order.item_total (item total)
+  // TODO: Probably this amount should be retrieved from the payments linked to the order
+  if (refundAmount && MathBN.gt(refundAmount, order.item_total)) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Refund amount cannot be greater than order total.`
+    )
+  }
+}
+
+function prepareFulfillmentData({
+  order,
+  input,
+  returnShippingOption,
+}: {
+  order: OrderDTO
+  input: OrderWorkflow.CreateOrderReturnWorkflowInput
+  returnShippingOption: {
+    id: string
+    provider_id: string
+    service_zone: { fulfillment_set: { location?: { id: string } } }
+  }
+}) {
+  const inputItems = input.items
+  const orderItemsMap = new Map<string, Required<OrderDTO>["items"][0]>(
+    order.items!.map((i) => [i.id, i])
+  )
+  const fulfillmentItems = inputItems.map((i) => {
+    const orderItem = orderItemsMap.get(i.id)!
+    return {
+      line_item_id: i.id,
+      quantity: i.quantity,
+      return_quantity: i.quantity,
+      title: orderItem.variant_title ?? orderItem.title,
+      sku: orderItem.variant_sku || "",
+      barcode: orderItem.variant_barcode || "",
+    } as FulfillmentWorkflow.CreateFulfillmentItemWorkflowDTO
+  })
+
+  let locationId: string | undefined = input.location_id
+  if (!locationId) {
+    locationId = returnShippingOption.service_zone.fulfillment_set.location?.id
+  }
+
+  if (!locationId) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `No location id provided and no stock location found in the fulfillment set of the shipping option with id ${returnShippingOption.id}.`
```

Comment:
suggestion(non-blocking): I would just do:
```suggestion
      `Cannot create return without stock location`
```

In theory, this shouldn't happen right? Because a shipping option can't live outside the context of a location.

Reply:
Yes it shouldn't, but I rather prevent a case where the user didn't create the link between the shipping option and the stock location wdyt?

Reply:
Ah yes, that makes sense